src/stream/src/executor/managed_state/aggregation/extreme.rs

```diff
@@ -123,261 +97,160 @@ pub trait ManagedTableState<S: StateStore>: Send + Sync + 'static {
-    ) -> StreamExecutorResult<Self> {
-        // Create the internal state based on the value we get.
-        Ok(Self {
-            top_n: BTreeMap::new(),
-            total_count: row_count,
-            _phantom_data: PhantomData::default(),
-            upstream_pk_len: pk_data_types.len(),
-            top_n_count,
+        pk_indices: PkIndices,
+        col_mapping: Arc<StateTableColumnMapping>,
+        row_count: usize,
+        cache_capacity: Option<usize>,
+    ) -> Self {
+        let group_key_len = group_key.map(|row| row.size()).unwrap_or(0);
+        // map agg column to cache row column index
+        let state_table_col_idx = col_mapping
+            .upstream_to_state_table(agg_call.args.val_indices()[0])
+            .expect("the column to be aggregate must appear in the state table");
+        let cache_agg_col_idx = state_table_col_idx - group_key_len;
+        // map order by columns to cache row column indices
+        let cache_order_pairs = [OrderPair::new(
+            cache_agg_col_idx,
+            match agg_call.kind {
+                AggKind::Min => OrderType::Ascending,
+                AggKind::Max => OrderType::Descending,
+                _ => unreachable!(),
+            },
+        )]
+        .into_iter()
+        .chain(pk_indices.iter().map(|idx| {
+            let state_table_col_idx = col_mapping
+                .upstream_to_state_table(*idx)
+                .expect("the pk columns must appear in the state table");
+            OrderPair::new(state_table_col_idx - group_key_len, OrderType::Ascending)
+        }))
+        .collect();
+
+        Self {
+            _phantom_data: PhantomData,
             group_key: group_key.cloned(),
-        })
+            state_table_col_mapping: col_mapping,
+            total_count: row_count,
+            cache: Cache::new(cache_capacity, cache_order_pairs),
+            cache_agg_col_idx,
+        }
     }
 
-    fn pk_length(&self) -> usize {
-        self.upstream_pk_len
+    fn group_key_len(&self) -> usize {
+        match &self.group_key {
+            Some(row) => row.size(),
+            None => 0,
+        }
     }
 
-    /// Retain only top n elements in the cache
-    fn retain_top_n(&mut self) {
-        if let Some(count) = self.top_n_count {
-            match EXTREME_TYPE {
-                variants::EXTREME_MIN => {
-                    while self.top_n.len() > count {
-                        self.top_n.pop_last();
-                    }
-                }
-                variants::EXTREME_MAX => {
-                    while self.top_n.len() > count {
-                        self.top_n.pop_first();
-                    }
-                }
-                _ => unimplemented!(),
-            }
-        }
+    fn state_row_to_cache_row(&self, state_row: &Row) -> Row {
+        state_row.by_indices(
+            &(0..self.state_table_col_mapping.len())
+                .skip(self.group_key_len())
+                .collect_vec(),
```

Comment:
How about storing this indices somewhere, or letting the `Row::by_indices` accept an `IntoIterator`?